Fix #20400, Start-Process with -NoNewWindow returns null ExitCode on Process object with -PassThru

CodeCyclone opened the pull request:
# PR Summary

Fixing issue #20400 along with some duplicates.

This issue was introduced by change #19096.
This change moved from .NET creating process objects to use P/Invoke (Windows APIs). The code would fetch the Process object from .NET but by design it throws an exception silently and ExitCode is null because it didn't handle spawning the process itself.

Calling the Handle property on the fetched Process will resolve ExitCode state because it fetches the underlying SafeHandle used to update the ExitCode property.

## PR Context

When you call Start-Process with -NoNewWindow, ExitCode will be null. This breaks a lot of customers of PowerShell both internally and externally. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

CodeCyclone committed:
Update Process.cs to fix #20400

Fixing issue #20400.

This issue was introduced by change #19096.
This change moved from .NET creating process objects to use P/Invoke (Windows APIs). The code would fetch the Process object from .NET but by design it throws an exception silently and ExitCode is null because it didn't spawn the process.

Calling Handle property on the fetched Process will resolve ExitCode state because it fetches the underlying SafeHandle used to update the ExitCode property.

iSazonov commented:
@CodeCyclone Please add test(s) to avoid regressions.

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2126,6 +2126,9 @@ protected override void BeginProcessing()
                     jobAssigned = jobObject.AssignProcessToJobObject(processInfo.Process);
                 }
 
+                // Fetch process handle to trigger Process object to update its state by calling SetProcessHandle, discard result.
```
Would be good to link to the issue so people looking at the code can see why this process handle is retrieved.

SteveL-MSFT reviewed (changes requested):
Please add a test in https://github.com/PowerShell/PowerShell/blob/811efa46df822bf7be6179b6219f9f9d160eb7d5/test/powershell/Modules/Microsoft.PowerShell.Management/Start-Process.Tests.ps1

CodeCyclone committed:
Improved comment, addressing PR feedback

CodeCyclone committed:
Added test

CodeCyclone commented:
Added test, addressed feedback on commenting change. Added more detail and referenced issue.
Screenshot of the test failing without my fix:
![image](https://github.com/PowerShell/PowerShell/assets/9702314/353172a9-e551-4112-8b6b-56a326ef4c81)

SteveL-MSFT reviewed (approved):
LGTM

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2126,6 +2126,12 @@ protected override void BeginProcessing()
                     jobAssigned = jobObject.AssignProcessToJobObject(processInfo.Process);
                 }
 
+                // Addresses issue #20400: Start-Process does not populate the process object's exit code with NoNewWindow switch
```
We never add links to issues.

CodeCyclone committed:
Update Process.cs

Removed referencing issue

daxian-dbw committed:
Update src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

Moazzem-Hossain-pixel reviewed (commented):
Thanks a lot
